keep/api/core/db.py

```diff
@@ -4583,37 +4583,49 @@ def get_last_alert_by_fingerprint(
 
 
 def set_last_alert(
-    tenant_id: str, alert: Alert, session: Optional[Session] = None
+    tenant_id: str, alert: Alert, session: Optional[Session] = None, max_retries=3
 ) -> None:
     logger.info(
         f"Set last alert for `{alert.fingerprint}`"
     )
     with existed_or_new_session(session) as session:
-        with session.begin_nested() as transaction:
-            last_alert = get_last_alert_by_fingerprint(tenant_id, alert.fingerprint, session, for_update=True)
+        for attempt in range(max_retries):
+            with session.begin_nested() as transaction:
+                try:
+                    last_alert = get_last_alert_by_fingerprint(tenant_id, alert.fingerprint, session, for_update=True)
 
-            # To prevent rare, but possible race condition
-            # For example if older alert failed to process
-            # and retried after new one
-            if last_alert and last_alert.timestamp.replace(tzinfo=tz.UTC) < alert.timestamp.replace(tzinfo=tz.UTC):
+                    # To prevent rare, but possible race condition
+                    # For example if older alert failed to process
+                    # and retried after new one
+                    if last_alert and last_alert.timestamp.replace(tzinfo=tz.UTC) < alert.timestamp.replace(tzinfo=tz.UTC):
 
-                logger.info(
-                    f"Update last alert for `{alert.fingerprint}`: {last_alert.alert_id} -> {alert.id}"
-                )
-                last_alert.timestamp = alert.timestamp
-                last_alert.alert_id = alert.id
-                session.add(last_alert)
+                        logger.info(
+                            f"Update last alert for `{alert.fingerprint}`: {last_alert.alert_id} -> {alert.id}"
+                        )
```

Comment:
wouldn't it print thousands of logs?

Reply:
LastAlert function has a lot of logs for now, for better transparency for the first time. I will clean them up after a while